gradle-plugin/src/main/java/com/squareup/anvil/plugin/AnvilPlugin.kt

```diff
@@ -145,29 +161,47 @@ open class AnvilPlugin : KotlinCompilerPluginSupportPlugin {
   @OptIn(ExperimentalStdlibApi::class)
   private fun disableIncrementalKotlinCompilation(
     project: Project,
-    isAndroidProject: Boolean
+    isAndroidProject: Boolean,
+    extension: AnvilExtension
   ) {
-    project.tasks
-        .withType(KaptGenerateStubsTask::class.java)
-        .configureEach { stubsTask ->
-          // Disable incremental compilation for the stub generating task. Trigger the compiler
-          // plugin if any dependencies in the compile classpath have changed. This will make sure
-          // that we pick up any change from a dependency when merging all the classes. Without
-          // this workaround we could make changes in any library, but these changes wouldn't be
-          // contributed to the Dagger graph, because incremental compilation tricked us.
-          stubsTask.doFirst {
-            stubsTask.incremental = false
-            stubsTask.logger.info(
-                "Anvil: Incremental compilation enabled: ${stubsTask.incremental} (stub)"
-            )
-          }
-        }
-
     // Use this signal to share state between DisableIncrementalCompilationTask and the Kotlin
     // compile task. If the plugin classpath changed, then DisableIncrementalCompilationTask sets
     // the signal to false.
     @Suppress("UnstableApiUsage")
     val incrementalSignal = IncrementalSignal.registerIfAbsent(project)
+
+    if (extension.generateDaggerFactoriesOnly) {
+      // We don't need to disable the incremental compilation for the stub generating task, when we
+      // only generate Dagger factories. That's only needed for merging Dagger modules.
+      if (isAndroidProject) {
+        project.androidVariantsConfigure { variant ->
+          val compileTaskName = "kaptGenerateStubs${variant.name.capitalize(US)}Kotlin"
+          disableIncrementalCompilationAction(project, incrementalSignal, compileTaskName)
+        }
+      } else {
+        disableIncrementalCompilationAction(project, incrementalSignal, "kaptGenerateStubsKotlin")
+        disableIncrementalCompilationAction(
+            project, incrementalSignal, "kaptGenerateStubsTestKotlin"
+        )
+      }
+    } else {
+      project.tasks
+          .withType(KaptGenerateStubsTask::class.java)
+          .configureEach { stubsTask ->
+            // Disable incremental compilation for the stub generating task. Trigger the compiler
+            // plugin if any dependencies in the compile classpath have changed. This will make sure
+            // that we pick up any change from a dependency when merging all the classes. Without
+            // this workaround we could make changes in any library, but these changes wouldn't be
+            // contributed to the Dagger graph, because incremental compilation tricked us.
+            stubsTask.doFirst {
+              stubsTask.incremental = false
```

Comment:
I believe this might also mess with build caching, why not do it in configureEach?

Reply:
That's the main reason I change this at execution time and not config time. I don't want to invalidate the config cache and it doesn't need to. When the task runs, because something has changed in the classpath, then the task should not be incremental. If the task doesn't, because nothing has changed, then it doesn't matter whether this flag is true or false. It's really not a big deal. 

I also want to toggle this flag as late as possible, because I've seen the Kotlin Gradle Plugin to change this flag again. Anvil needs to have the last say.